### PR TITLE
test: `send-anywhere` pfm scenarios

### LINKED ISSUE #9966: well-known interchain denoms should be registered as vbank assets in multichain-testing environment
## What is the Problem Being Solved?

As an orchestration developer, I should be able to use well-known interchain assets in zoe offers as a Give or Want and a Payment.

## Description of the Design

```[tasklist]
### Dependencies
- [ ] https://github.com/Agoric/agoric-sdk/issues/9752
```

- The `multichain-testing` environment should support offers with the native tokens of chains running in CI (`uatom`, `uosmo`, `uion`).  This likely requires: 
  - updates to `multichain-testing/scripts/fetch-starship-chain-info.ts`, to query the observed denoms of the environment
  - a proposal that does something along the lines of `publishInterchainAssetFromBank`, as part of `proposals/revise-chain-info.js` or a separate proposal
  
- Existing denom mappings, like `ATOM` -> `ibc/toyatom` should be overridden to resolve to the correct denom

### LINKED ISSUE #10445: local-orchestration-account: transfer supports multi-hop pfm routes
## What is the Problem Being Solved?

As a FastUSDC developer, I expect `LocalOrchestration.transfer(destination: ChainAddress: amount: AmountArg)` to support multi-hop PFM routes.

Example: transfer 1 uusdc from agoric to osmosis routed through noble.

See #10006 for the parent ticket.

### PR DESCRIPTION
**Pull request by 0xpatrickdev**
closes: #9966
closes: #10445

## Description
Adds multi-hop (PFM) scenarios to the `examples/send-anywhere.contract.js` multichain (e2e) test.  To support this change, this PR also includes:
  - a proposal for registering interchain assets in vbank (closes #9966). aims for production quality but is only used in tests 
  - a `fundFaucet` helper in `multichain-testing` so developers can request ATOM, OSMO, etc in `provisionSmartWallet`
  - a `GoDuration` type in `@agoric/orchestration` that captures basic Go [time duration strings](https://pkg.go.dev/time#ParseDuration) and an update to `DefaultPfmTimeoutOpts` (10min -> 10m)

### Security Considerations

`@agoric/builders/scripts/testing/register-interchain-bank-assets.js` allows callers overwrite assets in `vbank` and `agoricNames`. It's only intended for testing, and shouldn't be used in production. A production version might guard against accidental overrides.

### Scaling Considerations
None, mostly test code. Adds a little CI time to `multichain-testing` for the extra CoreEval.

### Documentation Considerations
None

### Testing Considerations
Includes an E2E to test in `multichain-testing` that leverages `register-interchain-bank-assets.js`. Also includes the first E2E test of PFM functionality added in #10584 and #10571.

### Upgrade Considerations
None, library code an NPM orch or FUSDC release.

### CODE REVIEW

File: multichain-testing/Makefile [hunk 1]
```diff
@@ -79,6 +79,11 @@ override-chain-registry:
 	node_modules/.bin/tsx scripts/fetch-starship-chain-info.ts && \
 	node_modules/.bin/tsx scripts/deploy-cli.ts src/revise-chain-info.builder.js
 
+register-bank-assets:
+	node_modules/.bin/tsx scripts/fetch-starship-chain-info.ts && \
+	node_modules/.bin/tsx scripts/deploy-cli.ts src/register-interchain-bank-assets.builder.js \
+		assets="$$(node_modules/.bin/tsx scripts/make-bank-asset-info.ts)"
+
```
Comment (dckc) [inline multichain-testing/Makefile]: what does it cost to get the scripts to know how to execute themselves?

```suggestion
register-bank-assets:
	./scripts/fetch-starship-chain-info.ts && \
	./scripts/deploy-cli.ts src/register-interchain-bank-assets.builder.js \
		assets="$$(./scripts/make-bank-asset-info.ts)"

```

File: multichain-testing/Makefile [hunk 2]
```diff
@@ -101,5 +106,5 @@ wait-for-pods:
 	node_modules/.bin/tsx scripts/pod-readiness.ts
 
 .PHONY: start
-start: install wait-for-pods port-forward fund-provision-pool override-chain-registry
+start: install wait-for-pods port-forward fund-provision-pool override-chain-registry register-bank-assets
```
Comment (dckc) [inline multichain-testing/Makefile:109]: Does `deploy-cli.ts` know how to run multiple evals in one chain governance proposal? How about combining `override-chain-registry` and  `register-bank-assets`?

File: multichain-testing/scripts/make-bank-asset-info.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+/* eslint-env node */
+
+import '@endo/init';
+import starshipChainInfo from '../starship-chain-info.js';
+import { makeAssetInfo } from '../tools/asset-info.ts';
+
+const main = () => {
+  if (!starshipChainInfo) {
+    throw new Error(
+      'starshipChainInfo not found. run `make override-chain-registry` or `node_modules/.bin/tsx scripts/fetch-starship-chain-info.ts` first.',
```
Comment (dckc) [inline multichain-testing/scripts/make-bank-asset-info.ts]: the `node_modules/.bin/` again jumps out at me... it looks like we're peeking into the internals of something.

```suggestion
      'starshipChainInfo not found. run `make override-chain-registry` or `./scripts/fetch-starship-chain-info.ts` first.',
```

File: multichain-testing/tools/faucet-tools.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+     */
+    fundFaucet: async (assets: [ChainName, Denom][], qty = DEFAULT_QTY) => {
+      const faucetAddr = agd.keys.showAddress('faucet');
+      console.log(`Faucet address: ${faucetAddr}`);
+
+      const fundAndTransfer = makeFundAndTransfer(
+        t,
+        retryUntilCondition,
+        useChain,
+      );
```
Comment (dckc) [inline multichain-testing/tools/faucet-tools.ts]: hoist this above `return`?

File: packages/builders/scripts/testing/register-interchain-bank-assets.js [hunk 5]
```diff
@@ -0,0 +1,190 @@
+/**
+ * @file register-interchain-bank-assets.js
```
Comment (dckc) [inline packages/builders/scripts/testing/register-interchain-bank-assets.js]: This doesn't look like a builder script. It looks like on-chain core-eval code.

File: packages/builders/scripts/testing/register-interchain-bank-assets.js [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+    brand: { produce: produceBrands },
+    issuer: { produce: produceIssuers },
+    installation: {
+      consume: { mintHolder },
+    },
+  },
+  { options: { assets } },
+) => {
+  trace(`${publishInterchainAssets.name} starting...`);
+  trace(assets);
```
Comment (dckc) [inline packages/builders/scripts/testing/register-interchain-bank-assets.js:84]: Darn; hoped to see the output of these `trace(...)` calls in [ci logs](https://github.com/Agoric/agoric-sdk/actions/runs/12123427233/job/33798929261?pr=10591#step:12:137), but no joy.

We're starting a new `mintHolder` contract per asset, it seems. How many new contracts are we starting here?

I suppose it doesn't matter, for testing purposes. But it reminds me of an `interchainMints.js` prototype.
https://github.com/agoric-labs/ag-power-tools/pull/8#issuecomment-2049734725

Comment (dckc) [reply packages/builders/scripts/testing/register-interchain-bank-assets.js:84]: Also, I'm reminded that cosgov's "add vault collateral" feature assumes the asset is not already in vbank. So if we add to vbank using a mechanism like this and then later want to add it as a vault collateral, we would need to enhance cosgov to handle that.

File: packages/builders/scripts/testing/register-interchain-bank-assets.js [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+    /**
+     * `addAssetToVault.js` will register the issuer with the `reserveKit`,
+     * but we don't need to do that here.
+     */
+
+    await Promise.all([
+      E(E(agoricNamesAdmin).lookupAdmin('issuer')).update(issuerName, issuer),
+      E(E(agoricNamesAdmin).lookupAdmin('brand')).update(issuerName, brand),
+      // triggers benign UnhandledPromiseRejection 'Error: keyword "ATOM" must
+      // be unique' in provisionPool in testing environments
```
Comment (dckc) [inline packages/builders/scripts/testing/register-interchain-bank-assets.js:138]: I wonder if the provisionPool could/should check that there's a relevant PSM before calling `addIssuer(...)`.

File: packages/builders/scripts/testing/register-interchain-bank-assets.js [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+    produceBrands[keyword].reset();
+    produceIssuers[keyword].reset();
+    produceBrands[keyword].resolve(brand);
+    produceIssuers[keyword].resolve(issuer);
+
+    // publish brand info / boardAux for offer legibility
+    await publishBrandInfo(
+      // @ts-expect-error 'Promise<StorageNode | null>' is not assignable to
+      // parameter of type 'ERef<StorageNode>'
+      chainStorage,
```
Comment (dckc) [inline packages/builders/scripts/testing/register-interchain-bank-assets.js:152]: boring boring boring... I wonder if we should just change the `BootstrapPowers` type for `chainStorage`.

File: packages/orchestration/src/exos/chain-hub.js [hunk 9]
```diff
@@ -179,7 +179,7 @@ const ChainIdArgShape = M.or(
 const DefaultPfmTimeoutOpts = harden(
   /** @type {const} */ ({
     retries: 3,
-    timeout: '10min',
+    timeout: /** @type {GoDuration} */ ('10m'),
```
Comment (dckc) [inline packages/orchestration/src/exos/chain-hub.js]: odd... `@type {const}` isn't enough?

File: packages/orchestration/src/proposals/start-send-anywhere.js [hunk 10]
```diff
@@ -82,13 +81,32 @@ export const startSendAnywhere = async (
     }),
   );
 
+  const safeLookup = async p =>
```
Comment (dckc) [inline packages/orchestration/src/proposals/start-send-anywhere.js]: the name `safeLookup` suggests that the `.lookup(...)` call should be in this function. `safeFulfill` maybe?

File: multichain-testing/src/register-interchain-bank-assets.builder.js [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+    values: { assets },
+  } = parseArgs({
+    args: scriptArgs,
+    options: parserOpts,
+  });
+
+  const parseAssets = () => {
+    if (typeof assets !== 'string') {
+      throw Error(
+        'must provide --assets=JSON.stringify({ denom: Denom; keyword: string; decimalPlaces: number; }[])',
```
Comment (dckc) [inline multichain-testing/src/register-interchain-bank-assets.builder.js]: Is this really a keyword? Or is it a misnomer where we really mean `issuerName`?

```suggestion
        'must provide --assets=JSON.stringify({ denom: Denom; issuerName: string; decimalPlaces: number; }[])',
```

File: multichain-testing/test/send-anywhere.test.ts [hunk 12]
```diff
@@ -123,12 +131,12 @@ const sendAnywhereScenario = test.macro({
...
     }
   },
 });
 
-test.serial(sendAnywhereScenario, 'osmosis', 1);
-test.serial(sendAnywhereScenario, 'osmosis', 2);
-test.serial(sendAnywhereScenario, 'cosmoshub', 1);
-test.serial(sendAnywhereScenario, 'cosmoshub', 2);
+test.serial(sendAnywhereScenario, 'osmosis', 1, 'IST');
+test.serial(sendAnywhereScenario, 'osmosis', 2, 'ATOM'); // exercises PFM (agoric -> cosmoshub -> osmosis)
```
Comment (dckc) [inline multichain-testing/test/send-anywhere.test.ts:140]: it took me a while to grok what this test does, but I think I get it now.

For ref:
https://github.com/Agoric/agoric-sdk/actions/runs/12123427233/job/33798929261?pr=10591#step:13:3141

File: multichain-testing/tools/ibc-transfer.ts [hunk 13]
```diff
@@ -118,14 +119,14 @@ export const makeIBCTransferMsg = (
 };
 
 export const createFundedWalletAndClient = async (
-  t: ExecutionContext,
+  log: (...args: unknown[]) => void,
   chainName: string,
   useChain: MultichainRegistry['useChain'],
 ) => {
   const { chain, creditFromFaucet, getRpcEndpoint } = useChain(chainName);
   const wallet = await createWallet(chain.bech32_prefix);
```
Comment (dckc) [inline multichain-testing/tools/ibc-transfer.ts]: `createWallet(...)` seems to generate random key material using ambient authority.

ah... ok; it takes key material as a default arg, a la...
 - #10038 

I'd rather see this test take advantage of that and pass in the key material, but I suppose we'll get there in due course.

**Review by dckc — APPROVED** [review]
looks good.

the comments here are mostly thinking-out-loud.

Putting `register-interchain-bank-assets.js` under `packages/builders/scripts` seems pretty odd; it's not a script. So I hope you'll move that. But I can still get behind this if you don't.

File: packages/orchestration/src/exos/chain-hub.js (hunk 9 above)
Comment (0xpatrickdev) [reply packages/orchestration/src/exos/chain-hub.js]: Good catch `const` should be sufficient here

File: multichain-testing/Makefile (hunk 1 above)
Comment (0xpatrickdev) [reply multichain-testing/Makefile]: I'll see where we land in #10605 and incorporate that approach

File: multichain-testing/Makefile (hunk 2 above)
Comment (0xpatrickdev) [reply multichain-testing/Makefile:109]: Good idea. I won't tackle that here but I'll hope to make this update soon

File: multichain-testing/src/register-interchain-bank-assets.builder.js (hunk 11 above)
Comment (0xpatrickdev) [reply multichain-testing/src/register-interchain-bank-assets.builder.js]: I remember this comment in the add-vault proposal. I'll look to make the corresponding change in the builder so we don't continue to carry the tech debt

File: packages/builders/scripts/testing/register-interchain-bank-assets.js (hunk 5 above)
Comment (0xpatrickdev) [reply packages/builders/scripts/testing/register-interchain-bank-assets.js]: I adjusted the @file comment to mention this. Are you suggesting this should live in a different directory?

File: multichain-testing/tools/ibc-transfer.ts (hunk 13 above)
Comment (0xpatrickdev) [reply multichain-testing/tools/ibc-transfer.ts]: Agree, but none of these tests rely on a specific key currently. For now, I've made `mnemonic` an optional parameter to `createFundedWalletAndClient`.

File: multichain-testing/Makefile (hunk 1 above)
Comment (dckc) [reply multichain-testing/Makefile]: yeah; don't worry about it for this PR

File: packages/builders/scripts/testing/register-interchain-bank-assets.js (hunk 5 above)
Comment (dckc) [reply packages/builders/scripts/testing/register-interchain-bank-assets.js]: yes. it doesn't seem to belong in `builder/scripts/`

not a hill I'd die on.

**Pull request merged**